pkg/ccl/testccl/sqlccl/sqlccl_test: TestActiveCancelSession failed

pkg/ccl/testccl/sqlccl/sqlccl_test.TestActiveCancelSession [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15119722?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15119722?buildTab=artifacts#/) on release-23.2.5-rc @ [57ab10786d6ed277959c8990e2823772893ac873](https://github.com/cockroachdb/cockroach/commits/57ab10786d6ed277959c8990e2823772893ac873):


```
=== RUN   TestActiveCancelSession
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/5c751b11a9f469cbdd173f9eebde1a6d/logTestActiveCancelSession2269752954
    test_log_scope.go:81: use -show-logs to present logs inline
    run_control_test.go:289: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5c751b11a9f469cbdd173f9eebde1a6d/logTestActiveCancelSession2269752954
--- FAIL: TestActiveCancelSession (3.07s)
=== RUN   TestActiveCancelSession/Tenant
    run_control_test.go:224: session not canceled; actual error: read tcp 127.0.0.1:36498->127.0.0.1:38811: read: connection reset by peer
    --- FAIL: TestActiveCancelSession/Tenant (0.03s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestActiveCancelSession.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-38444

--- a/pkg/ccl/testccl/sqlccl/run_control_test.go
+++ b/pkg/ccl/testccl/sqlccl/run_control_test.go
@@ -18,6 +18,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqltestutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -205,6 +206,27 @@ func testCancelSession(t *testing.T, hasActiveSession bool) {
 					_, err = conn1.ExecContext(ctx, "SELECT pg_sleep(1000000)")
 					errChan <- err
 				}()
+				// Block until the query goroutine was spun up and began
+				// executing the query - this is needed to avoid a race between
+				// canceling the session before vs after 'pg_sleep' query begins
+				// (the former would result in an unexpected error message).
+				testutils.SucceedsSoon(t, func() error {
+					row := conn2.QueryRowContext(ctx, `
+SELECT count(*) FROM [SHOW CLUSTER QUERIES] WHERE query LIKE '%pg_sleep%'
+                                              AND query NOT LIKE '%SHOW CLUSTER QUERIES%'
+;`)
+					var count int
+					if err = row.Scan(&count); err != nil {
+						t.Fatal(err)
+					}
+					if count == 1 {
+						return nil
+					}
+					if count > 1 {
+						t.Fatalf("unexpectedly found %d pg_sleep queries", count)
+					}
+					return errors.New("pg_sleep query hasn't started yet")
+				})
 			}
 
 			// Cancel the session on node 1.

